### DISCUSSION TITLE
Merge page from same document source

### DISCUSSION THREAD

**Discussion by loust333** [gettalong/hexapdf#312] [Q&A] [answered]
Hello @gettalong, 

I have the following use case: 

- @pdf_origin is the main pdf "main.pdf" 
- I fill the @pdf_origin with the data required 
- I need to add more pages from "main.pdf" page 3 to "main.pdf" (merge same document source)

Solution that I invented:

1. Fill main.pdf
2. Fill page 3 with the fields available 
3. Extra data for page 3 is filled in temp_pdf_origin 
4. Page 3 from temp_pdf_origin is insert in page 3 from @pdf_origin 

Problem: 

The page 3 from temp_pdf_origin is not correctly inserted in the @pdf_origin but the same values from @pdf_origin page 3 are inserted in the new concatenated page.

MAIN PDF: [Formulaire.AN.RD.pdf](https://github.com/user-attachments/files/16132496/Formulaire.AN.RD.pdf)

RESULT page 3 same as page 4 : [test.pdf](https://github.com/user-attachments/files/16132471/test.pdf) 

```
# frozen_string_literal: true

class PdfMergerService
  def initialize(source:, data: {})
    @source = "public/#{source}"
    @pdf_origin = HexaPDF::Document.open(@source)
    @pdf_origin.acro_form.fill(data)
    @pdf_origin.acro_form.recalculate_fields
  end

  def merge(temp_source: nil, pages: [])
    temp_source ||= @source

    pages.each do |page|
      puts page[:data], page[:from_index]
      @pdf_origin.pages.insert(page[:to_index], insert_page(temp_source:, data: page[:data], index: page[:from_index]))
    end
  end

  def read_io
    io = StringIO.new
    @pdf_origin.write(io)
    io.rewind
    io
  end

  private

  def insert_page(temp_source:, data:, index:)
    temp_pdf_origin = HexaPDF::Document.open(temp_source)
    temp_pdf_origin.acro_form.fill(data)
    temp_pdf_origin.pages[index]
  end
end



pdf_merger = PdfMergerService.new(source: "Formulaire.AN.RD.pdf", data: benefits_information_presenter.data)
pdf_merger.merge(pages: benefits_information_presenter.extra_car_pages)

```

I also tried with the merge example from the website and got the same result..

**Comment by gettalong** [comment, marked as answer]
Hello @loust333,

There are a few problems with the code:

* The main reason why this doesn't work is that AcroForm fields with the same full name are deemed to be just one field. When you copy the page, the field names stay the same and thus page 4 doesn't have separate fields.

* The way you copy the page in `#merge` doesn't work in the general case. You need to correctly import the page (see https://hexapdf.gettalong.org/examples/merging.html#merging-pdf-files for details), otherwise there *will* be problems.

* AcroForm fields are actually separate from their appearance on a page. This allows, for example, a single field to appear in multiple places. So when you want to copy a page from a PDF *and* the underlying form fields, just copying the page is not enough. The form fields also need to be registered in the main AcroForm dictionary so that they are recognized.

There is a todo item on my list for making it easier to copy pages with form fields to another PDF. I have bumped up its priority to make that happen sooner than later.

**Reply by loust333** [reply]
Hello @gettalong, 

Thanks for the explanation. Should I try to use the merge and update the dictionary, or should I wait for the TODO to be completed ?

**Reply by gettalong** [reply]
There are a few things to do to merge AcroForm data, e.g. making sure the full field names are unique and that the default appearance strings are set correctly. So, alas, it is not as straight-forward.

Another possible way: If you don't need the resulting PDFs to be editable afterwards, you could just fill the data in and then flatten the form fields, i.e. in `#insert_page` after you fill in the data you would call `temp_pdf_origin.acro_form.flatten`.

**Reply by loust333** [reply]
Hey @gettalong,

I created a temporary solution based of your feedback and the discussion https://github.com/gettalong/hexapdf/discussions/250 and the documentation :

```ruby
class PdfMergerService
  def initialize(source:, data: {})
    @source = "public/#{source}"
    @pdf_origin = HexaPDF::Document.open(@source)
    @pdf_origin.acro_form.fill(data)
    @pdf_origin.acro_form.recalculate_fields
  end

  def merge(temp_source: nil, pages: [])
    temp_source ||= @source

    i = 1
    pages.reverse_each do |page|
      temp_pdf_origin = HexaPDF::Document.open(temp_source)
      temp_pdf_origin.acro_form.fill(page[:data])
      copy_page = HexaPDF::Importer.for(temp_pdf_origin).import(temp_pdf_origin.pages[page[:from_index]])

      fields = copy_page[:Annots].select { |annot| annot[:Subtype] == :Widget && annot[:T].present? }
      fields.each do |field|
        field[:T] = "#{field[:T]}_additional_page_#{i}"
      end

      @pdf_origin.pages.insert(page[:to_index], @pdf_origin.import(copy_page))
      i += 1
    end
  end

  def read_io
    io = StringIO.new
    @pdf_origin.write(io)
    io.rewind
    io
  end
end
```
Result: [result.pdf](https://github.com/user-attachments/files/16164194/result.pdf)

**Reply by gettalong** [reply]
Nice!

However, I think you can just replace the `copy_page = ...` line with `copy_page = @pdf_origin.import(temp_pdf_origin.pages[page[:from_index]])` since you are creating a separate `HexaPDF::Document` object in each iteration (which is different to the case in the linked discussion). Then you don't need to import the page twice.

And what you are missing is registering the fields with the main AcroForm dictionary, i.e. `@pdf_origin.acro_form[:Fields] << field` inside the `fields.each` block. I'm actually not quite sure in which cases this is actually needed since most viewers display the form fields correctly anyway but just to be on the safe side.

**Reply by loust333** [reply]
I updated the code and it works without any issues.  👍
